File: Python/2-Medium/00003._无重复字符的最长子串/00003._无重复字符的最长子串.py
def length_of_longest_substring(s: str) -> int:
    """
    寻找给定字符串中不含重复字符的最长子串长度。

    参数:
        s (str): 输入的字符串

    返回:
        int: 最长子串的长度

    示例:
        >>> length_of_longest_substring("abcabcbb")
        3
        >>> length_of_longest_substring("bbbbb")
        1
    """
    # 初始化指针和哈希表
    start, max_length = 0, 0
    char_index_map = {}

    for end in range(len(s)):
        if s[end] in char_index_map and char_index_map[s[end]] >= start:
            # 更新start指针，防止向回移动
            start = char_index_map[s[end]] + 1
        # 更新字符位置和最大长度
        char_index_map[s[end]] = end
        max_length = max(max_length, end - start + 1)

    return max_length

# 运行代码并打印重要信息
test_strings = ["abcabcbb", "bbbbb", "", "b", "pwwkew"]
results = {test: length_of_longest_substring(test) for test in test_strings}
print(results)

File: C++/2-Medium/00003._无重复字符的最长子串/00003._无重复字符的最长子串.cpp

#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> charIndexMap; // 存储字符和其最新索引的映射
    int maxLength = 0; // 最长子串长度
    for (int start = 0, end = 0; end < s.length(); end++) {
        if (charIndexMap.find(s[end]) != charIndexMap.end()) {
            // 更新start指针，排除重复字符
            start = max(start, charIndexMap[s[end]] + 1);
        }
        // 更新字符的最新索引
        charIndexMap[s[end]] = end;
        // 更新最长子串长度
        maxLength = max(maxLength, end - start + 1);
    }
    return maxLength;
}

int main() {
    string input = "abcabcbb";
    cout << "Input: " << input << "\\n";
    cout << "Output: " << lengthOfLongestSubstring(input) << "\\n";
    return 0;
}